### PR TITLE
Implement signature importing

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Implement the logic for `IMetaDataEmit::TranslateSigWithScope` (and `IMetaDataEmit::ImportTypeDef`) in `dnmd::interfaces`.

As this logic is really complex, and is the only code that needs to do signature walking, I wanted to implement it in a separate PR from the `IMetaDataEmit` PR to help reduce the size and complexity of that one.

This PR still needs tests. I think I'll finally add a "make a new Metadata image" API to `dnmd` (in a separate PR) to enable us to add unit tests for bespoke scenarios (the majority of the code in this PR is bespoke scenarios that will be difficult to test by only using existing assemblies).

### CODE REVIEW

File: src/inc/dnmd.hpp [hunk 1]
```diff
@@ -16,4 +16,41 @@ struct mdhandle_deleter_t final
 // C++ lifetime wrapper for mdhandle_t type
 using mdhandle_ptr = std::unique_ptr<mdhandle_t, mdhandle_deleter_t>;
 
+struct md_added_row_t
+{
+private:
+    mdcursor_t new_row;
+public:
+    md_added_row_t() = default;
+    md_added_row_t(mdcursor_t row) : new_row(row) {}
```
Comment (AaronRobinsonMSFT) [inline src/inc/dnmd.hpp:25]:
```suggestion
    md_added_row_t(mdcursor_t row) : new_row{ row } {}
```

File: src/inc/dnmd.hpp [hunk 2]
```diff
@@ -16,4 +16,41 @@ struct mdhandle_deleter_t final
 // C++ lifetime wrapper for mdhandle_t type
 using mdhandle_ptr = std::unique_ptr<mdhandle_t, mdhandle_deleter_t>;
 
+struct md_added_row_t
```
Comment (AaronRobinsonMSFT) [inline src/inc/dnmd.hpp:19]:
```suggestion
struct md_added_row_t final
```

File: src/inc/dnmd.hpp [hunk 3]
```diff
@@ -16,4 +16,41 @@ struct mdhandle_deleter_t final
...
+    md_added_row_t(md_added_row_t&& other)
+    {
+        *this = std::move(other);
+    }
+
+    md_added_row_t& operator=(md_added_row_t const& other) = delete;
+    md_added_row_t& operator=(md_added_row_t&& other)
+    {
+        new_row = other.new_row;
+        new (&other.new_row) mdcursor_t(); // Clear the other's row so we don't double-commit.
```
Comment (AaronRobinsonMSFT) [inline src/inc/dnmd.hpp:36]: This requires placement new, which I would like to avoid.

Instead how about the following:

```c++
other.new_row = {};
```

File: src/inc/internal/span.hpp [hunk 4]
```diff
@@ -47,6 +47,31 @@ class span
...
+    {
+        return { _ptr, _size };
+    }
+
+    T* begin() noexcept
+    {
+        return _ptr;
+    }
+
+    T* begin() const noexcept
```
Comment (AaronRobinsonMSFT) [inline src/inc/internal/span.hpp:61]:
```suggestion
    T const* cbegin() const noexcept
```

File: src/inc/internal/span.hpp [hunk 5]
```diff
@@ -47,6 +47,31 @@ class span
...
+    {
+        return _ptr;
+    }
+
+    T* end() noexcept
+    {
+        return _ptr + _size;
+    }
+
+    T* end() const noexcept
```
Comment (AaronRobinsonMSFT) [inline src/inc/internal/span.hpp:71]:
```suggestion
    T const* cend() const noexcept
```

File: src/inc/internal/span.hpp [hunk 6]
```diff
@@ -47,6 +47,31 @@ class span
             throw std::out_of_range{ "Out of bounds access" };
         return _ptr[idx];
     }
+
+    operator span<const T>() const
```
Comment (AaronRobinsonMSFT) [inline src/inc/internal/span.hpp:51]:
```suggestion
    operator span<T const>() const
```

File: src/inc/internal/span.hpp [hunk 7]
```diff
@@ -89,6 +114,11 @@ class owning_span final : public span<T>
         this->_ptr = {};
         return tmp;
     }
+
+    operator owning_span<const T, Deleter>() const
```
Comment (AaronRobinsonMSFT) [inline src/inc/internal/span.hpp:118]:
```suggestion
    operator owning_span<T const, Deleter>() const
```

File: src/interfaces/importhelpers.hpp [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+#define _SRC_INTERFACES_IMPORTHELPERS_HPP
+
+#include <internal/dnmd_platform.hpp>
+#include <internal/span.hpp>
+#include <functional>
+
+HRESULT ImportReferenceToTypeDef(
+    mdcursor_t sourceTypeDef,
+    mdhandle_t sourceAssembly,
+    span<const uint8_t> sourceAssemblyHash,
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/importhelpers.hpp:11]:
```suggestion
    span<uint8_t const> sourceAssemblyHash,
```

File: src/interfaces/importhelpers.hpp [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    mdhandle_t targetAssembly,
+    mdhandle_t targetModule,
+    bool alwaysImport,
+    std::function<void(mdcursor_t row)> onRowEdited,
+    mdcursor_t* targetTypeDef);
+
+HRESULT ImportReferenceToTypeDefOrRefOrSpec(
+    mdhandle_t sourceAssembly,
+    mdhandle_t sourceModule,
+    span<const uint8_t> sourceAssemblyHash,
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/importhelpers.hpp:21]:
```suggestion
    span<uint8_t const> sourceAssemblyHash,
```

File: src/interfaces/pal.cpp [hunk 10]
```diff
@@ -130,3 +139,80 @@ int strcat_s(char* dest, rsize_t destsz, char const* src)
     return 0;
 }
 #endif // !defined(__STDC_LIB_EXT1__) && !defined(BUILD_WINDOWS)
+
+#if defined(BUILD_WINDOWS)
+namespace
+{
+    struct BCRYPT_ALG_HANDLE_deleter
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/pal.cpp:146]:
```suggestion
    struct BCRYPT_ALG_HANDLE_deleter final
```

File: src/interfaces/pal.cpp [hunk 11]
```diff
@@ -130,3 +139,80 @@ int strcat_s(char* dest, rsize_t destsz, char const* src)
...
+    {
+        void operator()(BCRYPT_ALG_HANDLE h) const noexcept
+        {
+            ::BCryptCloseAlgorithmProvider(h, 0);
+        }
+    };
+
+    using bcrypt_alg_handle = std::unique_ptr<void, BCRYPT_ALG_HANDLE_deleter>;
+
+    struct BCRYPT_HASH_HANDLE_deleter
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/pal.cpp:156]:
```suggestion
    struct BCRYPT_HASH_HANDLE_deleter final
```

File: src/interfaces/pal.hpp [hunk 12]
```diff
@@ -107,6 +109,10 @@ namespace pal
             return _ptr;
         }
     };
+
+    constexpr size_t SHA1_HASH_SIZE = 20;
+
+    bool ComputeSha1Hash(span<const uint8_t> data, std::array<uint8_t, SHA1_HASH_SIZE>& hashDestination);
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/pal.hpp:115]:
```suggestion
    bool ComputeSha1Hash(span<uint8_t const> data, std::array<uint8_t, SHA1_HASH_SIZE>& hashDestination);
```

File: src/interfaces/signatures.cpp [hunk 13]
```diff
@@ -1,25 +1,30 @@
...
         signature = slice(signature, CorSigUncompressData(signature, &value));
         return std::make_tuple(value, signature);
     }
 
-    std::tuple<int32_t, span<uint8_t>> read_compressed_int(span<uint8_t> signature)
+    template<typename T, typename = typename std::enable_if<std::is_same<typename std::remove_const<T>::type, uint8_t>::value>::type>
+    std::tuple<int32_t, span<T>> read_compressed_int(span<T> signature)
     {
-        int value;
+        int value = 0;
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/signatures.cpp:21]: This one is initialized but not the others, is this an oversight?

**Review by AaronRobinsonMSFT — COMMENTED** [review]
There is a lot going on in this PR. I need to do another pass.

File: src/inc/dnmd.hpp (hunk 3 above)
Comment (jkoritzinsky) [reply src/inc/dnmd.hpp:36]:
```suggestion
        other.new_row = {}; // Clear the other's row so we don't double-commit.
```

File: src/interfaces/signatures.cpp (hunk 13 above)
Comment (jkoritzinsky) [reply src/interfaces/signatures.cpp:21]: I'll check, but it believe this case required the initialization because of how `CorSigUncompressSignedInt` handles the sign calculations was leading to an "uninitialized value" warning. It's possible that the other cases have the same problem but the compiler only understood this case.

File: src/interfaces/importhelpers.cpp [hunk 14]
```diff
@@ -0,0 +1,1639 @@
...
+
+            StrongNameToken asmPublicKeyToken;
+            RETURN_IF_FAILED(StrongNameTokenFromPublicKey({ publicKey, publicKeyLength }, asmPublicKeyToken));
+
+            if (refPublicKeyOrTokenLength != asmPublicKeyToken.size() || !std::equal(asmPublicKeyToken.begin(), asmPublicKeyToken.end(), refPublicKeyOrToken))
+                return S_FALSE;
+            
+            return S_OK;
+        }
+        return E_NOTIMPL;
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/importhelpers.cpp:1112]: Is this correct?

File: src/interfaces/importhelpers.cpp [hunk 15]
```diff
@@ -0,0 +1,1639 @@
...
+        
+        mdhandle_t sourceModule = md_extract_handle_from_cursor(sourceTypeRef);
+        mdguid_t targetModuleMvid = { 0 };
+        mdguid_t targetAssemblyMvid = { 0 };
+        mdguid_t sourceAssemblyMvid = { 0 };
+        mdguid_t sourceModuleMvid = { 0 };
+        RETURN_IF_FAILED(GetMvid(targetModule, &targetModuleMvid));
+        RETURN_IF_FAILED(GetMvid(targetAssembly, &targetAssemblyMvid));
+        RETURN_IF_FAILED(GetMvid(sourceAssembly, &sourceAssemblyMvid));
+        RETURN_IF_FAILED(GetMvid(sourceModule, &sourceModuleMvid));
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/importhelpers.cpp:1149]:
```suggestion
        mdguid_t targetModuleMvid = {};
        mdguid_t targetAssemblyMvid = {};
        mdguid_t sourceAssemblyMvid = {};
        mdguid_t sourceModuleMvid = {};
        RETURN_IF_FAILED(GetMvid(targetModule, &targetModuleMvid));
        RETURN_IF_FAILED(GetMvid(targetAssembly, &targetAssemblyMvid));
        RETURN_IF_FAILED(GetMvid(sourceModule, &sourceModuleMvid));
        RETURN_IF_FAILED(GetMvid(sourceAssembly, &sourceAssemblyMvid));
```

File: src/interfaces/importhelpers.cpp [hunk 16]
```diff
@@ -0,0 +1,1639 @@
...
+        bool sameModuleMvid = std::memcmp(&targetModuleMvid, &sourceModuleMvid, sizeof(mdguid_t)) == 0;
+        bool sameAssemblyMvid = std::memcmp(&targetAssemblyMvid, &sourceAssemblyMvid, sizeof(mdguid_t)) == 0;
+
+        // II.22.38 1. Valid ResolutionScope values
+        // - null
+        // - TypeRef token
+        // - ModuleRef token
+        // - Module token
+        // - AssemblyRef token
+        mdcursor_t targetOutermostScope = { 0 };
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/importhelpers.cpp:1160]:
```suggestion
        mdcursor_t targetOutermostScope = {};
```

File: src/interfaces/importhelpers.cpp [hunk 17]
```diff
@@ -0,0 +1,1639 @@
...
+                return E_FAIL;
+
+            if (IsNilToken(scopeToken))
+            {
+                // A Nil ResolutionScope means a reference to an ExportedType entry
+                // in the assembly.
+                // Since the source and target assemblies have the same identity,
+                // we can use the Nil token and we don't have to resolve the ExportedType
+                // as the target and source assemblies are the same.
+                targetOutermostScope = { 0 };
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/importhelpers.cpp:1185]:
```suggestion
                targetOutermostScope = {};
```

File: src/interfaces/importhelpers.cpp [hunk 18]
```diff
@@ -0,0 +1,1639 @@
...
+            if (!md_cursor_to_token(scope, &scopeToken))
+                return E_FAIL;
+            
+            if (IsNilToken(scopeToken))
+            {
+                // Lookup ExportedType entry in the source assembly for this type.
+                mdcursor_t exportedType;
+                uint32_t count;
+                bool foundExportedType = false;
+                mdcursor_t implementation = { 0 };
```
Comment (AaronRobinsonMSFT) [inline src/interfaces/importhelpers.cpp:1266]:
```suggestion
                mdcursor_t implementation = {};
```

File: src/interfaces/signatures.cpp [hunk 19]
```diff
@@ -278,3 +286,243 @@ malloc_span<std::uint8_t> GetMethodDefSigFromMethodRefSig(span<uint8_t> methodRe
...
 }
+
+// Define a function object that enables us to combine multiple lambdas into a single overload set.
+namespace
+{
+    template<typename... T>
+    struct Overload;
+
+    template<typename T>
+    struct Overload<T>
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/interfaces/signatures.cpp:297]: I've used this pattern before and seen it around the web for years. I think the one I pulled this from was either https://stackoverflow.com/a/18432618 or one similar to that.

I generally only use this pattern when I have tag-types.

File: src/interfaces/importhelpers.cpp [hunk 20]
```diff
@@ -0,0 +1,1639 @@
...
+                // Handle the case when a ref may be using a full public key instead of a token.
+                StrongNameToken refPublicKeyToken;
+
+                uint32_t assemblyRefFlags;
+                if (1 != md_get_column_value_as_constant(c, mdtAssemblyRef_Flags, 1, &assemblyRefFlags))
+                    return CLDB_E_FILE_CORRUPT;
+                
+                if (IsAfPublicKey(flags) == IsAfPublicKey(assemblyRefFlags))
+                {
+                    // If both the source and destination either have a full key or a token, we can compare them directly.
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/interfaces/importhelpers.cpp:361]: I'll rephrase this to make it clearer.

File: src/interfaces/importhelpers.cpp [hunk 21]
```diff
@@ -0,0 +1,1639 @@
...
+                    if (tempBlobLength != publicKeyOrToken.size() || !std::equal(publicKeyOrToken.begin(), publicKeyOrToken.end(), tempBlob))
+                        continue;
+                }
+                else if (IsAfPublicKey(assemblyRefFlags))
+                {
+                    // This AssemblyRef row has a full public key. And our source has a token.
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/interfaces/importhelpers.cpp:367]: If the `if ((publicKeyOrToken.size() == 0) != (tempBlobLength == 0))` doesn't fail, then the source must have either a key or token. If `IsAfPublicKey(assemblyRefFlags)` is, then our assembly ref must have a public key. However, we already know by here that `IsAfPublicKey(flags) == IsAfPublicKey(assemblyRefFlags)` is not true, so our source must not have a full key. Therefore, it has a token. The `else` case is the same concept.

File: src/interfaces/importhelpers.cpp [hunk 22]
```diff
@@ -0,0 +1,1639 @@
...
+        // relating to using a sample compiler from the .NET Framework SDK from before VS6.0.
+        // This tool never shipped, so we don't need to account for this bug here.
+        return E_INVALIDARG;
+    }
+    else
+    {
+        RETURN_IF_FAILED(ImportReferenceToAssembly(sourceAssembly, sourceAssemblyHash, targetModule, targetAssembly, onRowAdded, &resolutionScope));
+    }
+
+    std::stack<mdcursor_t> typesForTypeRefs;
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/interfaces/importhelpers.cpp:744]: I'll add handling for memory allocation failures here.

**Review by AaronRobinsonMSFT — APPROVED** [review]
